cpp/arcticdb/column_store/key_segment.cpp

```diff
@@ -0,0 +1,146 @@
+    auto creation_ts_it = creation_ts_data.template cbegin<int64_TDT>();
+    auto content_hash_it = content_hash_data.template cbegin<uint64_TDT>();
+    auto key_types_it = key_types_data.template cbegin<uint8_TDT>();
+
+    if (symbol_structure_ == SymbolStructure::SAME && !is_sequence_type(start_indexes_->type().data_type())) {
+        // Can return AtomKeyPacked vector, much more time and memory efficient, and most common case
+        std::vector<AtomKeyPacked> res;
+        res.reserve(num_keys_);
+        auto index_start_it = index_start_data.template cbegin<int64_TDT>();
+        auto index_end_it = index_end_data.template cbegin<int64_TDT>();
+        for (size_t row_idx = 0;
+             row_idx < num_keys_;
+             ++row_idx, ++version_it, ++creation_ts_it, ++content_hash_it, ++index_start_it, ++index_end_it, ++key_types_it) {
+            res.emplace_back(*version_it, *creation_ts_it, *content_hash_it, KeyType(*key_types_it), *index_start_it, *index_end_it);
+        }
+        return res;
+    } else {
+        // Fall back to returning fully materialised AtomKeys
+        std::vector<AtomKey> res;
+        res.reserve(num_keys_);
+        auto stream_id_data = stream_ids_->data();
+        auto stream_id_it = stream_id_data.template cbegin<uint64_TDT>();
+        const bool stream_id_is_sequence_type = is_sequence_type(stream_ids_->type().data_type());
+        const bool index_is_sequence_type = is_sequence_type(start_indexes_->type().data_type());
+        auto index_start_it = index_start_data.template cbegin<int64_TDT>();
+        auto index_end_it = index_start_data.template cbegin<int64_TDT>();
+        for (size_t row_idx = 0;
+             row_idx < num_keys_;
+             ++row_idx, ++stream_id_it, ++version_it, ++creation_ts_it, ++content_hash_it, ++index_start_it, ++index_end_it, ++key_types_it) {
+            // String columns are stored as uint64s (offsets into the string pool), but StringPool::get_const_view expects an int64
+            // Hence the static cast
+            auto symbol_value = symbol_.value_or(
+                    stream_id_is_sequence_type ? StreamId(StringId(string_pool_->get_const_view(static_cast<int64_t>(*stream_id_it)))) : StreamId(NumericId(*stream_id_it))
+            );
+            // String columns are stored as uint64s, whereas timestamp index start/end values are stored as int64s (nanoseconds since epoch)
+            // We could add another if statement above based on index_is_sequence_type, but as these types are the same width,
+            // we can just do some fancy casting to reinterpret the value in the iterator as a uint64.
+            // We then need to static cast back to int64 for the same reason as above RE StringPool::get_const_view
+            auto index_start_value = index_is_sequence_type ?
+                    IndexValue(StringId(string_pool_->get_const_view(static_cast<int64_t>(*reinterpret_cast<const uint64_t*>(&(*index_start_it)))))) :
```

Comment:
This seems similar to the stuff in stream_utils.hpp, like get_range_for_segment, is there something in there that could be used, or maybe a well-named function that does this could be added. Probably should all be in index_utils.hpp as well to be fair